internal/api/mfa.go

```diff
@@ -61,6 +69,74 @@ const (
 	QRCodeGenerationErrorMessage   = "Error generating QR Code"
 )
 
+func (a *API) enrollSMSFactor(w http.ResponseWriter, r *http.Request, user *models.User, session *models.Session, params *EnrollFactorParams) error {
+	ctx := r.Context()
+	config := a.config
+	db := a.db.WithContext(ctx)
+	if !config.MFA.SMS.EnrollEnabled {
+		return unprocessableEntityError(ErrorCodeMFAEnrollDisabled, "mfa enroll is disabled for sms")
+	}
+	if params.PhoneNumber == "" {
+		return badRequestError(ErrorCodeValidationFailed, "Phone number required to enroll SMS factor")
+	}
+
+	phone, err := validatePhone(params.PhoneNumber)
+	if err != nil {
+		return badRequestError(ErrorCodeValidationFailed, "Invalid phone number format (E.164 required)")
+	}
+	factors := user.Factors
+
+	factorCount := len(factors)
+	numVerifiedFactors := 0
+	if err := models.DeleteExpiredFactors(db, config.MFA.FactorExpiryDuration); err != nil {
+		return err
+	}
+
+	for _, factor := range factors {
+		if factor.IsVerified() {
+			numVerifiedFactors += 1
+		}
+	}
+
+	if factorCount >= int(config.MFA.MaxEnrolledFactors) {
+		return forbiddenError(ErrorCodeTooManyEnrolledMFAFactors, "Maximum number of verified factors reached, unenroll to continue")
```

Comment:
This is not a [HTTP Forbidden](https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/403) error as it has nothing to do about the authorization of the user, but rather is an [Unprocessable Entity](https://www.rfc-editor.org/rfc/rfc4918#section-11.2).

Please also adjust below.

Reply:
Will do that in a separate PR, opening now

Reply:
Done here: https://github.com/supabase/auth/pull/1676

Reply:
we really need to stop changing the http status codes just because it's the "wrong" one - we don't know how many users rely on checking it over using error codes here

Reply:
We explicitly established a policy about what status codes are used for what, and how they change. They should not be used explicitly, and there is very little likelihood that someone is checking for a 403 if there are too many enrolled factors.

This is clearly encoded in our error codes docs: https://github.com/supabase/supabase/pull/22793/files#diff-b916e67d4ea2d5fcbf41c433d25b9f6637104001f1bfcab6ba9a1fe8cec94652R92

Reply:
Furthermore 403 cannot ever be used to represent this state, which I missed in the error codes PR.

Reply:
We can discuss the status code change for TOTP (if any) on the other PR. This particular change on the SMS code path should be OK imo since it's specific to SMS - a user will have to add an SMS factor to run into this code path so it's unlikely that anyone depends on it.